### PR TITLE
swappable device model

### PR DESCRIPTION
**Pull request by Akay7**
- [x] add swapper
- [x] add tests for endpoints
- [x] add tests for queryset
- [x] update documentation
- [x] check locally without swapped models
- [x] check locally with swapped models

### CODE REVIEW

**Comment by xtrinch** [comment]
@henriklindgren was kind enough to provide us with a test example in https://github.com/xtrinch/fcm-django/pull/240, so perhaps we can use that to add additional tests for this PR

**Comment by Akay7** [comment]
@xtrinch example [here](https://github.com/xtrinch/fcm-django-web-demo/pull/32) . 
Let me know if you would like have any changes in this PR.

File: README.rst [hunk 1]
```diff
@@ -380,14 +380,47 @@ logins on the same device, you do not wish the old user to receive messages whil
...
+
 Python 3 support
 ----------------
 - ``fcm-django`` is fully compatible with Python 3.7+
 - for Python 3.6, use ``fcm-django < 2.0.0`` , because `firebase-admin with version 6 drop support of Python 3.6 <https://firebase.google.com/support/release-notes/admin/python#version_600_-_06_october_2022>`_
 
 Django version compatibility
 ----------------------------
-Compatible with Django versions 3.0+.
+Compatible with Django versions 3.2+.
```
Comment (xtrinch) [inline README.rst:423]: Is this needed? Swapper appears to be compatible with 2.2+

Comment (xtrinch) [inline README.rst:423]: Is this needed? Swapper appears to be compatible with Django 2+

File: README.rst [hunk 2]
```diff
@@ -380,14 +380,47 @@ logins on the same device, you do not wish the old user to receive messages whil
 Via DRF, any creation of device with an already existing registration ID will be transformed into an update.
 If done manually, you are responsible for deleting the old device entry.
 
+Using custom FCMDevice model
+----------------------------
+
+If there's a need to store additional information or change type of fields in the FCMDevice model.
+You could simple override this model. It could be fully your own model or you could inheriting from AbstractFCMDevice
```
Comment (xtrinch) [inline README.rst:387]: If they use fully their own model, how is that compatible with fcm-django's migrations?

File: README.rst [hunk 3]
```diff
@@ -380,14 +380,47 @@ logins on the same device, you do not wish the old user to receive messages whil
...
+
+.. code-block:: python
+
+    FCM_DJANGO_FCMDEVICE_MODEL = "your_app.CustomDevice"
+
+Note: This functionality based on `Swapper <https://pypi.org/project/swapper/>`_ that based on functionality 
+that allow to use a `custom User model <https://docs.djangoproject.com/en/4.2/topics/auth/customizing/#substituting-a-custom-user-model>`_.
+So this functionality have the same limitations. 
+The most is important limitation it is that is difficult to start out with a default (non-swapped) model 
+and then later to switch to a swapped implementation without doing some migration hacking.
```
Comment (xtrinch) [inline README.rst:414]: A line on whether there's any specialty with migrations if a custom model is used and how to kickstart it would be nice

File: setup.py [hunk 4]
```diff
@@ -31,7 +31,7 @@
         "fcm_django/migrations",
     ],
     python_requires=">=3.7",
-    install_requires=["firebase-admin>=5,<7", "Django"],
+    install_requires=["firebase-admin>=5,<7", "Django", "swapper"],
```
Comment (xtrinch) [inline setup.py:34]: If the version in requirements is pinned, should it be pinned here too?

File: tests/settings.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
+import dj_database_url
```
Comment (xtrinch) [inline tests/settings.py:1]: There now appears to be `testing_settings.py` and `settings.py` in the same directory, this should be somehow dealt with

File: tests/test_api.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+    assert response.status_code == 201
+    assert FCMDevice.objects.count() == devices_qty + 1
+    device = FCMDevice.objects.get(registration_id=registration_id)
+    assert device.type == DeviceType.WEB
+    assert device.registration_id == registration_id
+
+
+@pytest.mark.django_db
+def test_drf_endpoint_add_device(client, registration_id):
```
Comment (xtrinch) [inline tests/test_api.py:51]: It would be beneficial to have a separate file for DRF and a separate file for tastypie tests

File: pytest.ini [hunk 7]
```diff
@@ -0,0 +1,4 @@
+[pytest]
+django_find_project = false
```
Comment (xtrinch) [inline pytest.ini:2]: We now have pyproject.toml & pytests.ini, we should really only have one of these, please combine somehow

File: README.rst (hunk 1 above)
Comment (Akay7) [reply README.rst:423]: Django 2.0 supports Python 3.4, 3.5, 3.6, and 3.7.
And [there's no support for Django 2.0](https://www.djangoproject.com/download/) on official page.
I though that we wont support releases that aren't have official security updates it could be great to update docs. Ie we drop support of Python <=3.6

But I can rollback those changes.

Comment (xtrinch) [reply README.rst:423]: We will support whatever works with the current setup, we shouldn't be as optinionated as to drop support just because something does not have security updates :-)

Comment (Akay7) [reply README.rst:423]: If so why was support of python3.6 dropped?

Comment (xtrinch) [reply README.rst:423]: Because firebase-admin does not support python 3.6 in their latest release. I'm not saying we should make an effort to support django 2, we should merely not remove it from compatible packages if it works

File: README.rst (hunk 2 above)
Comment (Akay7) [reply README.rst:387]: So it's a completely different model. 
* fcm-django's migrations migrate FCMDevice model
* other models migrations migrate other models

File: README.rst (hunk 1 above)
Comment (Akay7) [reply README.rst:423]: > Because firebase-admin does not support python 3.6 in their latest release.
But that doesn't mean that it wont works with current version of package. They removed it because python 3.6 become deprecated.

File: setup.py (hunk 4 above)
Comment (Akay7) [reply setup.py:34]: We don't pin Django here and pin it in requirements. I though do the same with swapper.

File: README.rst (hunk 2 above)
Comment (xtrinch) [reply README.rst:387]: So they have an empty FCMDevice table in their DB because of that? This should be at least documented

File: README.rst (hunk 1 above)
Comment (xtrinch) [reply README.rst:423]: Unsupported by who?

File: README.rst (hunk 2 above)
Comment (Akay7) [reply README.rst:387]: But it's how does Django works during overriding `User` model. There's a package that use [swapper](https://pypi.org/project/django-cities/#swappable-models) and they don't copy paste Django or Swapper docs.

File: README.rst (hunk 3 above)
Comment (Akay7) [reply README.rst:414]: You could copy [branch](https://github.com/Akay7/fcm-django-web-demo/tree/swappedModels) to your repo and I could mention that diff between two branches in this repo.

File: tests/settings.py (hunk 5 above)
Comment (Akay7) [reply tests/settings.py:1]: So there's two attempts to run all tests against everything:
* with default settings and non swapped models `settings.py` 
* and with overwritten settings only that required by swapper - `testing_settings.py`

File: README.rst (hunk 2 above)
Comment (xtrinch) [reply README.rst:387]: I was not asking you to copy paste django or swapper docs, but to answer a common question that might arise from users in the documentation...

File: tests/swapped_models/migrations/0001_initial.py [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+                    ),
+                ),
+                (
+                    "date_created",
+                    models.DateTimeField(
+                        auto_now_add=True, null=True, verbose_name="Creation date"
+                    ),
+                ),
+                (
+                    "device_id",
```
[the remark this answers is not in the document]
Comment (Akay7) [reply tests/swapped_models/migrations/0001_initial.py:42]: It's migration for swapped model.

File: setup.py (hunk 4 above)
Comment (xtrinch) [reply setup.py:34]: Yeah. It's strange that firebase is pinned and django is not

File: tests/test_api.py (hunk 6 above)
Comment (Akay7) [reply tests/test_api.py:51]: there's not much to separate just a single test for tastypie

File: README.rst (hunk 3 above)
Comment (xtrinch) [reply README.rst:414]: I would rather not maintain two branches, maybe we can make the fcm-django-web-demo have two separate examples

File: tests/settings.py (hunk 5 above)
Comment (xtrinch) [reply tests/settings.py:1]: Two attempts, but three files?

https://github.com/Akay7/fcm-django/blob/swapableDeviceModel/tests/settings.py
https://github.com/Akay7/fcm-django/blob/swapableDeviceModel/tests/testing_settings.py
https://github.com/Akay7/fcm-django/blob/swapableDeviceModel/tests/swap_settings.py

File: tests/swapped_models/migrations/0001_initial.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (xtrinch) [reply tests/swapped_models/migrations/0001_initial.py:42]: If I am not 100% clear on what is happening in this PR, I will not be able to maintain it nor will anyone else. Please document the test setup, it shouldn't take you more than a couple of minutes to write a couple of lines...

File: README.rst (hunk 1 above)
Comment (Akay7) [reply README.rst:423]: > Unsupported by who?

by Django core team.

File: README.rst (hunk 3 above)
Comment (xtrinch) [reply README.rst:414]: For me personally it's way harder to maintain two branches. Someone opens an issue it needs to be fixed in two places in two separate branches, these two branches eventually cannot be reconciliated anymore, epecially if there's PRs involved

**Comment by Akay7** [comment]
so i have todo:

- [ ] 1. Move pytest to pyproject
- [ ] 2. Add docs about there will be two separate models if swapper used, and you have to control your Django migrations by yourself
- [ ] 3. Mention that Django 3.0 supported instead 3.2

File: README.rst (hunk 1 above)
Comment (Akay7) [reply README.rst]: I rollback this change

File: README.rst (hunk 3 above)
Comment (Akay7) [reply README.rst:414]: Okay I will make copy of project in the main branch.

File: tests/settings.py (hunk 5 above)
Comment (Akay7) [reply tests/settings.py:1]: Uh oh. Yeah, I removed extra.

File: tests/swapped_models/migrations/0001_initial.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Akay7) [reply tests/swapped_models/migrations/0001_initial.py:42]: I will add comment to this file.

File: README.rst (hunk 2 above)
Comment (Akay7) [reply README.rst:387]: > but to answer a common question that might arise from users in the documentation..

I updated doc with answers on questions that you raise

**Comment by xtrinch** [comment]
@henriklindgren / @olzhasar  If any of you have any extra time on your hands, maybe throw an eye on this PR 🙏

File: fcm_django/models.py [hunk 9]
```diff
@@ -387,3 +388,5 @@ class Meta:
         indexes = [
             models.Index(fields=["registration_id", "user"]),
         ]
+
+        swappable = swapper.swappable_setting("fcm_django", "fcmdevice")
```
[the remark this answers is not in the document]
Comment (olzhasar) [reply fcm_django/models.py:392]: I also believe that inheriting from `AbstractFCMDevice` should be mandatory in case one wants to use a custom Device model for us to ensure that logic is preserved. But in that case we should completely refactor our models. We probably should move all the fields from abstract models to `FCMDevice` model, and in abstract models we can use class attributes corresponding to fields that are used in business logic, similar to what Django does in User model. For our case this can look like this:

```python
class AbstractFCMDevice:
    TOKEN_FIELD: str

    @property
    def token(self):
        return getattr(self, self.TOKEN_FIELD)

    def send_message(self, message, app, **kwargs):
        message.token = self.token
        ...


class FCMDevice(AbstractFCMDevice):
    registration_id = models.CharField()

    TOKEN_FIELD = "registration_id"
```

File: tox.ini [hunk 10]
```diff
@@ -17,5 +22,12 @@ deps =
     -rrequirements_test.txt
     django32: Django>=3.2,<3.3
     django42: Django>=4.2,<4.3
+    postgres: psycopg2-binary>=2.9.6
```
Comment (olzhasar) [inline tox.ini:25]: There are so many unrelated things in this PR. Why do we need postgres and mariadb here? Even if we have reasons, that's definitely a topic for a separate discussion in a separate PR

**Review by olzhasar — CHANGES_REQUESTED** [review]
Overall, it feels like a lot more work needs to be done in order to make the solution complete

File: tests/settings.py [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+                "django.template.context_processors.debug",
+                "django.template.context_processors.request",
+                "django.contrib.auth.context_processors.auth",
+                "django.contrib.messages.context_processors.messages",
+            ],
+        },
+    },
+]
+
+IS_SWAP = False
```
[the remark this answers is not in the document]
Comment (Akay7) [reply tests/settings.py:38]: Yeah it could be. But in that case we have to introduce developer into how swapper works.

File: tests/swapped_models/models.py [hunk 12]
```diff
@@ -0,0 +1,22 @@
+import uuid
+
+from django.db import models
+
+from fcm_django.models import AbstractFCMDevice
+
+
+class CustomDevice(AbstractFCMDevice, models.Model):
```
[the remark this answers is not in the document]
Comment (Akay7) [reply tests/swapped_models/models.py]: Thanks, I have corrected it.

File: tox.ini (hunk 10 above)
Comment (Akay7) [reply tox.ini:25]: > Why do we need postgres and mariadb here?

It's used only for tests. 
1. There's been added functionality that could swap model. 
2. This project contain a lot of logic that tied to the model.
3. I wish there were tests that would test the functionality of the database, which are more strict than Sqlite

File: tests/test_api.py (hunk 6 above)
Comment (Akay7) [reply tests/test_api.py]: > Honestly, there is too much stuff happening in this PR.

Yep, that's true.

> Tests for api endpoints should be handled in a separate PR.

Integration of swapper could fail functionality for an API. There were no tests for an API so I added it.

> And I agree with @xtrinch that separate files are better cause we are testing completely unrelated concepts.

I split it into two files.

File: tests/test_models.py [hunk 13]
```diff
@@ -98,19 +102,42 @@ def test_firebase_error(
...
         with pytest.raises(FirebaseError, match=str(firebase_error)):
             fcm_device.send_message(message)
 
-        mock_fcm_device_deactivate.assert_called_once_with(
-            fcm_device, fcm_device.registration_id, firebase_error
+        fcm_device.refresh_from_db()
+        # device is still active because error is unknown
+        assert fcm_device.active
+
+    def test_firebase_invalid_registration_error(
```
[the remark this answers is not in the document]
Comment (Akay7) [reply tests/test_models.py:118]: Does it make sense to leave test writing to the future PR?

File: tox.ini (hunk 10 above)
Comment (olzhasar) [reply tox.ini:25]: Nothing in the functionality that was added in this PR requires `postgres` or `mariadb` for tests

File: tests/test_models.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (olzhasar) [reply tests/test_models.py:118]: It does. PRs should be about one thing. Otherwise they are much more expensive to polish, review and merge

File: fcm_django/models.py [hunk 14]
```diff
@@ -1,6 +1,7 @@
 from copy import copy
 from typing import List, NamedTuple, Sequence, Union
 
+import swapper
```
[the remark this answers is not in the document]
Comment (Akay7) [reply fcm_django/models.py:4]: > I don't see any reason why do we need this external dependency

I see the reason for this:
1. is that it is already done this way in [django cities](https://pypi.org/project/django-cities/)
2. that done for that in that for this package
3. it hasn't been done yet for that package in any other way yet

File: tests/settings.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (Akay7) [reply tests/settings.py:38]: If comment comment for this line of code make any sense i could add it.

But if your point that the same thing could be done in different way just because it could be done I don't think that is a constructive.

File: tests/test_models.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (Akay7) [reply tests/test_models.py:118]: > It does. PRs should be about one thing. Otherwise they are much more expensive to polish, review and merge

Is your point that tests should be delivered after changes that those tests should check?

_There are three way to write tests:_
1. _Before writing the code_
2. _During writing the code_
3. _Never(some people call it "later")_

File: tests/settings.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (olzhasar) [reply tests/settings.py:38]: You are making an API for users to consume. Why would you force your users to do additional steps if you can handle this on your side?

File: fcm_django/models.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (olzhasar) [reply fcm_django/models.py:4]: I do not understand your reasoning. Just because some other package does it, does not mean it's a good decision in the first place

File: tox.ini (hunk 10 above)
Comment (Akay7) [reply tox.ini:25]: > Nothing in the functionality that was added in this PR requires `postgres` or `mariadb` for tests

Will tests on more databases break anything?

SQLite doesn't enforce `max_length` for text fields. So test `test_registration_id_size` could never fail. 
But yep:
* it could be done in a different PR. 
* Should it done in a different pr? - Well, I don't think that it's required.

File: fcm_django/models.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (xtrinch) [reply fcm_django/models.py:392]: I also think inheriting from AbstractFCMDevice should be mandatory. Otherwise the entire purpose of this package is void and it breaks logic

[the remark this answers is not in the document]
Comment (Akay7) [reply fcm_django/models.py:392]: > I also think inheriting from AbstractFCMDevice should be mandatory. Otherwise the entire purpose of this package is void and it breaks logic

What should be done for that? Mention in the doc that _CustomDevice_ should be always inherited from `AbstractFCMDevice`?

[the remark this answers is not in the document]
Comment (xtrinch) [reply fcm_django/models.py:392]: At the very least that yeah 🤔

File: fcm_django/models.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (xtrinch) [reply fcm_django/models.py:4]: Is it upgradeable later? If so, I'm not really set on any fixed solution. Anything goes that accomplishes the job and can be changed later

File: fcm_django/models.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (xtrinch) [reply fcm_django/models.py:392]: I would remove that line, yep

File: tox.ini (hunk 10 above)
Comment (Akay7) [reply tox.ini:25]: > As I do think this PR contains significant changes for this repo, it would greatly enhance review-ability if you could extract the non-swappable-related changes into a separate PR, if you would be up for it.

@xtrinch 

Make two PRs:
1. With tests of existed functionality;
2. With swapper functionality and tests for it?

Or do you have a different view of the breakdown of these PRs?

Comment (xtrinch) [reply tox.ini:25]: No, sounds like a good plan

File: tests/settings.py [hunk 15]
```diff
@@ -37,4 +34,5 @@
         },
     },
 ]
-FIREBASE_APP = initialize_app()
```
Comment (Akay7) [inline tests/settings.py:40]: Somehow initialization of `FIREBASE_APP` isn't required for test case scenarios.

File: fcm_django/models.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (Akay7) [reply fcm_django/models.py:4]: > Is it upgradeable later?

If we want to go the way @olzhasar suggested, we can always copy the swapper inside this project.

File: fcm_django/models.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Akay7) [reply fcm_django/models.py:392]: > I would remove that line, yep

I have changed those lines in the docs.

File: tox.ini [hunk 16]
```diff
@@ -9,8 +9,6 @@ python =
     3.7: py37
     3.8: py38
     3.9: py39
-    3.9: py39
```
Comment (Akay7) [inline tox.ini:12]: I removed the duplicates

File: tests/conftest.py [hunk 17]
```diff
@@ -59,11 +62,3 @@ def mock_firebase_send(mocker: MockerFixture, firebase_message_id_send):
     mock = mocker.patch("fcm_django.models.messaging.send")
     mock.return_value = firebase_message_id_send
     return mock
-
-
-@pytest.fixture
-def mock_fcm_device_deactivate(mocker: MockerFixture):
-    return mocker.patch(
-        "fcm_django.models.FCMDevice.deactivate_devices_with_error_result",
```
Comment (Akay7) [inline tests/conftest.py:67]: Since the model could be Swapped, this fixture may not work. So I removed it and corrected a few tests.

File: docs/index.rst [hunk 18]
```diff
@@ -201,6 +201,14 @@ Sending messages in bulk
     # Or (send_message parameters include: messages, dry_run, app)
     FCMDevice.objects.send_message(Message(...))
 
+    Sending messages raises all the errors that ``firebase-admin`` raises, so make sure
```
Comment (Akay7) [inline docs/index.rst:204]: That part of docs was missed from `index.rst` So I just synchronize content of `docs/index.rst` and `README.rst`

**Comment by Akay7** [comment]
@xtrinch If there is expected anything else than let me know.

File: tests/settings.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (xtrinch) [reply tests/settings.py:38]: If it's just for tests, then it does not matter.

File: fcm_django/models.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (xtrinch) [reply fcm_django/models.py:392]: There is one more thing from this conversation to pull out -> we should skip migrating FCMDevice, if the user has decided to overwrite the model. There is no reason for the user to have two tables, one that they do not use

[the remark this answers is not in the document]
Comment (Akay7) [reply fcm_django/models.py:393]: > There is one more thing from this conversation to pull out -> we should skip migrating FCMDevice, if the user has decided to overwrite the model. There is no reason for the user to have two tables, one that they do not use

In order to accomplish that user don't have to register `fcm_django` at `INSTALLED_APPS`. I updated docs and settings for tests.

File: tests/swapped_models/models.py [hunk 19]
```diff
@@ -0,0 +1,22 @@
...
+
+
+class CustomDevice(AbstractFCMDevice):
+    # fields could be overwritten
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
+    # NOTE: the max_length could not be supported or enforced on the database level
+    # https://docs.djangoproject.com/en/4.2/ref/models/fields/#django.db.models.CharField.max_length
+    registration_id = models.CharField(unique=True, max_length=515)
+    # could be added new fields
+    more_data = models.TextField()
```
[the remark this answers is not in the document]
Comment (Akay7) [reply tests/swapped_models/models.py:15]: I added two tests:
* `test_fields_on_the_device_can_be_redefined_by_swapped_model`
* `test_fields_on_the_device_can_be_added_by_swapped_model`

File: tests/settings.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (Akay7) [reply tests/settings.py]: I added a comment to this line so that there would be no confusion.

File: fcm_django/models.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (xtrinch) [reply fcm_django/models.py:393]: I'm not 100% sure about this solution to be honest. This will be problematic if we ever add any other models. Not saying we will, but you know.. What do you think @olzhasar

[the remark this answers is not in the document]
Comment (Akay7) [reply fcm_django/models.py:393]: > This will be problematic if we ever add any other models. Not saying we will, but you know.

@xtrinch 
If there will be more models and few of them is required it could be done in the same way as it done at [allauth](https://django-allauth.readthedocs.io/en/latest/installation.html). Basically there are an app with nested apps. So to use fcm-django with integrated Device model user will have to add into installed apps:
```python
"fcm_django",
"fcm_django.device",
```

If user would like to use only required models out of `fcm_django` then adds only 
```python
"fcm_django"
```

And these changes can be made in the future, at this stage no modifications are required as there is only one model in the application.

[the remark this answers is not in the document]
Comment (olzhasar) [reply fcm_django/models.py:393]: @xtrinch I do agree that migrations for our default device model will become obsolete if one decides to use a custom model, but it's probably not that critical. There are other a lot more concerning issues that this PR entails.

[the remark this answers is not in the document]
Comment (xtrinch) [reply fcm_django/models.py:393]: @olzhasar Yeah, especially if it can be skipped by skipping it in installed_apps. Is there anything besides the comments you already left?

[the remark this answers is not in the document]
Comment (olzhasar) [reply fcm_django/models.py:393]: @xtrinch My biggest concern is still about introducing external dependency. I did not receive any valid reasons for doing that. I've just created a sample pull request to demonstrate how can we implement `swappable` behavior without using `swapper`. Here is a one particular commit you should see:
https://github.com/xtrinch/fcm-django/pull/248/commits/66039eaef7c30a2de6950ae42653acf72529950f

[the remark this answers is not in the document]
Comment (olzhasar) [reply fcm_django/models.py:393]: I also think that we should redesign abstract models to prepare everything for this PR. One of the most common reasons why one might be interested in using a custom Device model is because they don't need some fields that standard model provides. I don't see how the current implementation helps with that. Let's say I don't want `device_id` in my model. This field is not involved in any business logic and it's a reasonable wish. However, the only way for me to exclude it from my table (correct me if I'm wrong) is to set it to `None` in my custom model, e.g.:

```py3
class CustomDevice(AbstractFCMDevice):
    `device_id` = None
```

This way of doing feels completely unnatural. As a user, I would instead like `fcm-django` to keep only necessary fields in the abstract model and allow me to handle the rest.

Another concern is the admin interface. If I want to customize it, I will probably need to deregister the old admin and register a new one. This feels odd as well.

[the remark this answers is not in the document]
Comment (Akay7) [reply fcm_django/models.py:393]: > I've just created a sample pull request to demonstrate how can we implement `swappable` behavior without using `swapper`. Here is a one particular commit you should see

Your code 60 lines without doc strings and tests. Swapper contain 100 lines of code with doc strings and also contains tests. 

> One of the most common reasons why one might be interested in using a custom Device model is because they don't need some fields that standard model provides.

But in this PR it was your concern that inheritance from AbstractDevice should be mandatory. If there's no inheritance and user would like to have fully customized model then there's no problem.

> Another concern is the admin interface. If I want to customize it, I will probably need to deregister the old admin and register a new one. This feels odd as well.

Same as with customization User at Django itself.

[the remark this answers is not in the document]
Comment (xtrinch) [reply fcm_django/models.py:393]: I have received many requests over the years to be able to add fields to FCMDevice, never to remove them. I think removing fields is actually an edge case that may break something if the wrong field is chosen

[the remark this answers is not in the document]
Comment (olzhasar) [reply fcm_django/models.py:393]: @Akay7 Which 60 lines are you referring to? The replacement of `swapper` in my code is just a single function that is 14 lines of code. We can add docstrings and tests to this function prior to merging. The code is just to show you the idea that we don't any external library for our simple use case

[the remark this answers is not in the document]
Comment (olzhasar) [reply fcm_django/models.py:393]: @xtrinch Yes, removing fields should not be allowed, that's the reason why I think that abstract model should contain only necessary fields.
I used `fcm-django` in one project of mine and I didn't need `device_id` field there. So, not being forced to add it to my Custom Device model would have been great.

Also, I wonder what is the motivation for having two abstract models: `Device` and `FCMDevice`?

[the remark this answers is not in the document]
Comment (Akay7) [reply fcm_django/models.py:393]: > Which 60 lines are you referring to? The replacement of `swapper` in my code is just a single function that is 14 lines of code.

@olzhasar okay, my bad it's 14 lines. 
And there will be 14 lines for two models or it will be 28? Yeah, it's possible to implement swapper on our own but after reaching a certain complexity of the project, it will look about the same as the swapper. But that would also take up our time.

**Comment by Akay7** [comment]
@xtrinch , @olzhasar Is there a decision to be made? Or are we not moving in that direction with swapper yet?

**Comment by xtrinch** [comment]
I don't mind using swapper, overall it looks fine. But since this PR can have so many unintended consequences and unintended issues, I'm reluctant to just merge this and I would probably go with releasing this as an alpha version first

**Comment by Akay7** [comment]
> this PR can have so many unintended consequences and unintended issues

I can't imagine any of that problem.
In this PR most of the code is documentation and changes to run all tests twice once with swapped models and once with original.

> I'm reluctant to just merge this and I would probably go with releasing this as an alpha version first

I'm okay with release it as alpha. 

@xtrinch Please tell me if there's anything you need me to do.

**Review by xtrinch — APPROVED** [review]
